File: src/litellm_relay/proxy.py
from __future__ import annotations

import asyncio
from datetime import datetime, timezone
import json
from uuid import uuid4

from .config import RelayConfig, is_notion_host, normalize_host
from .pac import build_pac
from .shadow import ShadowClient


class RelayProxy:
    def __init__(self, config: RelayConfig):
        self.config = config
        self.shadow_client = ShadowClient(config)

    async def serve_forever(self) -> None:
        self.config.log_path.parent.mkdir(parents=True, exist_ok=True)
        server = await asyncio.start_server(
            self._handle_client, self.config.host, self.config.port
        )
        self._log(
            {
                "event": "relay_started",
                "listen": f"{self.config.host}:{self.config.port}",
                "shadow_enabled": self.config.shadow_enabled,
            }
        )
        async with server:
            await server.serve_forever()

    async def _handle_client(
        self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter
    ) -> None:
        peer = writer.get_extra_info("peername")
        try:
            header = await reader.readuntil(b"\r\n\r\n")
        except asyncio.IncompleteReadError:
            writer.close()
            await writer.wait_closed()
            return

        try:
            header_text = header.decode("iso-8859-1")
            request_line = header_text.splitlines()[0]
            method, target, _version = request_line.split(" ", 2)
        except (UnicodeDecodeError, ValueError, IndexError):
            await self._write_response(writer, 400, b"bad request\n")
            return

        if method.upper() == "GET" and target in {"/proxy.pac", "/pac"}:
            pac = build_pac(self.config).encode("utf-8")
            await self._write_response(
                writer, 200, pac, content_type="application/x-ns-proxy-autoconfig"
            )
            return

        if method.upper() == "CONNECT":
            await self._handle_connect(target, peer, reader, writer)
            return

        await self._write_response(
            writer,
            501,
            b"litellm-relay v0 only supports CONNECT tunneling and /proxy.pac\n",
        )

    async def _handle_connect(
        self,
        target: str,
        peer: object,
        client_reader: asyncio.StreamReader,
        client_writer: asyncio.StreamWriter,
    ) -> None:
        host, port = parse_connect_target(target)
        event = self._event(
            {
                "event": "connect",
                "method": "CONNECT",
                "host": host,
                "port": port,
                "peer": repr(peer),
                "notion_match": is_notion_host(host, self.config),
            }
        )

        if event["notion_match"]:
            shadow_result = self.shadow_client.maybe_shadow(event)
            event["shadow"] = {
                "attempted": shadow_result.attempted,
                "ok": shadow_result.ok,
                "status": shadow_result.status,
                "error": shadow_result.error,
                "event_id": shadow_result.event_id,
            }
        self._log(event)

        try:
            upstream_reader, upstream_writer = await asyncio.open_connection(host, port)
        except OSError as exc:
            self._log(
                self._event(
                    {
                        "event": "connect_failed",
                        "host": host,
                        "port": port,
                        "error": exc.__class__.__name__,
                    }
                )
            )
            await self._write_response(client_writer, 502, b"upstream connect failed\n")
            return

        client_writer.write(b"HTTP/1.1 200 Connection Established\r\n\r\n")
        await client_writer.drain()
        await tunnel(client_reader, client_writer, upstream_reader, upstream_writer)

    async def _write_response(
        self,
        writer: asyncio.StreamWriter,
        status: int,
        body: bytes,
        content_type: str = "text/plain",
    ) -> None:
        reason = {
            200: "OK",
            400: "Bad Request",
            501: "Not Implemented",
            502: "Bad Gateway",
        }.get(status, "OK")
        writer.write(
            (
                f"HTTP/1.1 {status} {reason}\r\n"
                f"content-length: {len(body)}\r\n"
                f"content-type: {content_type}\r\n"
                "connection: close\r\n"
                "\r\n"
            ).encode("ascii")
            + body
        )
        await writer.drain()
        writer.close()
        await writer.wait_closed()

    def _event(self, values: dict[str, object]) -> dict[str, object]:
        return {
            "event_id": str(uuid4()),
            "timestamp": datetime.now(timezone.utc).isoformat(),
            **values,
        }

    def _log(self, event: dict[str, object]) -> None:
        self.config.log_path.parent.mkdir(parents=True, exist_ok=True)
        with self.config.log_path.open("a", encoding="utf-8") as log_file:
            log_file.write(json.dumps(redact_event(event), sort_keys=True) + "\n")


async def tunnel(
    client_reader: asyncio.StreamReader,
    client_writer: asyncio.StreamWriter,
    upstream_reader: asyncio.StreamReader,
    upstream_writer: asyncio.StreamWriter,
) -> None:
    async def pipe(reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
        try:
            while chunk := await reader.read(65536):
                writer.write(chunk)
                await writer.drain()
        except (ConnectionError, asyncio.CancelledError):
            pass
        finally:
            writer.close()

    await asyncio.gather(
        pipe(client_reader, upstream_writer),
        pipe(upstream_reader, client_writer),
        return_exceptions=True,
    )


def parse_connect_target(target: str) -> tuple[str, int]:
    if target.startswith("["):
        host, _, rest = target[1:].partition("]")
        port = int(rest.lstrip(":") or "443")
        return host, port
    host, sep, port = target.partition(":")
    return normalize_host(host), int(port) if sep else 443


def redact_event(event: dict[str, object]) -> dict[str, object]:
    redacted = dict(event)
    for key in ("authorization", "cookie", "token_v2", "prompt", "body", "url"):
        redacted.pop(key, None)
    return redacted

File: src/proxy.rs
use std::{fs, net::SocketAddr, sync::Arc, time::Instant};

use anyhow::{Context, Result};
use chrono::Utc;
use rustls::pki_types::ServerName;
use serde_json::{json, Value};
use tokio::{
    io::AsyncWriteExt,
    net::{TcpListener, TcpStream},
};
use tokio_rustls::{TlsAcceptor, TlsConnector};
use uuid::Uuid;

use crate::{
    cert::{client_tls_config, ensure_ca, server_tls_config},
    config::{classify_host, is_ai_host, is_notion_host, RelayConfig},
    events::{append_event, clear_events, read_events},
    gateway::{CaptureIngest, GatewayClient, IngestResult},
    http::{
        build_http_payload, copy_bidirectional_counted, parse_connect_target, parse_limit,
        parse_request_line, parse_response_status, parse_route, parse_start_line,
        read_http_message, read_until_headers, redact_headers, should_close, write_response,
    },
    pac::build_pac,
};

const DASHBOARD_HTML: &str = include_str!("static/dashboard.html");

pub struct RelayProxy {
    config: Arc<RelayConfig>,
    gateway: GatewayClient,
}

impl RelayProxy {
    pub fn new(config: RelayConfig) -> Self {
        let config = Arc::new(config);
        let gateway = GatewayClient::new(Arc::clone(&config));
        Self { config, gateway }
    }

    pub async fn serve_forever(self) -> Result<()> {
        if let Some(parent) = self.config.log_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let listen = format!("{}:{}", self.config.host, self.config.port);
        let listener = TcpListener::bind(&listen).await?;
        self.log_event(json!({
            "event": "relay_started",
            "listen": listen,
            "capture_payloads": self.config.mitm_enabled,
            "shadow_enabled": self.config.shadow_enabled,
            "runtime": "rust",
        }))?;

        let proxy = Arc::new(self);
        loop {
            let (stream, peer) = listener.accept().await?;
            let proxy = Arc::clone(&proxy);
            tokio::spawn(async move {
                if let Err(error) = proxy.handle_client(stream, peer).await {
                    let _ = proxy.log_event(json!({
                        "event": "client_error",
                        "peer": peer.to_string(),
                        "error": error.to_string(),
                    }));
                }
            });
        }
    }

    async fn handle_client(&self, mut stream: TcpStream, peer: SocketAddr) -> Result<()> {
        let header = match read_until_headers(&mut stream).await {
            Ok(header) => header,
            Err(_) => return Ok(()),
        };
        let header_text = String::from_utf8_lossy(&header).to_string();
        let (method, target) = parse_start_line(&header_text)?;
        let route = parse_route(&target);

        if matches!(method.as_str(), "GET" | "HEAD")
            && matches!(route.path.as_str(), "/" | "/index.html")
        {
            return write_response(
                &mut stream,
                200,
                "text/html; charset=utf-8",
                DASHBOARD_HTML.as_bytes(),
                method == "GET",
            )
            .await;
        }

        if matches!(method.as_str(), "GET" | "HEAD")
            && matches!(route.path.as_str(), "/proxy.pac" | "/pac")
        {
            let pac = build_pac(&self.config);
            return write_response(
                &mut stream,
                200,
                "application/x-ns-proxy-autoconfig",
                pac.as_bytes(),
                method == "GET",
            )
            .await;
        }

        if method == "GET" && route.path == "/api/status" {
            return self.write_json(&mut stream, self.status_payload()?).await;
        }

        if method == "GET" && route.path == "/api/events" {
            let limit = parse_limit(&route.query);
            return self
                .write_json(
                    &mut stream,
                    json!({
                        "events": read_events(&self.config.log_path, limit),
                        "limit": limit,
                    }),
                )
                .await;
        }

        if method == "POST" && route.path == "/api/events/clear" {
            clear_events(&self.config.log_path)?;
            self.log_event(json!({
                "event": "relay_log_cleared",
                "listen": format!("{}:{}", self.config.host, self.config.port),
            }))?;
            return self.write_json(&mut stream, json!({"ok": true})).await;
        }

        if method == "CONNECT" {
            return self.handle_connect(stream, target, peer).await;
        }

        write_response(
            &mut stream,
            501,
            "text/plain",
            b"litellm-relay only supports CONNECT tunneling and local dashboard endpoints\n",
            true,
        )
        .await
    }

    async fn handle_connect(
        &self,
        mut client: TcpStream,
        target: String,
        peer: SocketAddr,
    ) -> Result<()> {
        let target = parse_connect_target(&target)?;
        let started_at = Instant::now();
        let event_id = Uuid::new_v4().to_string();
        let app = classify_host(&target.host, &self.config);
        let ai_match = is_ai_host(&target.host, &self.config);
        let notion_match = is_notion_host(&target.host, &self.config);
        let mut event = json!({
            "event_id": event_id,
            "event": "connect",
            "method": "CONNECT",
            "host": target.host,
            "port": target.port,
            "peer": peer.to_string(),
            "app": app,
            "ai_match": ai_match,
            "notion_match": notion_match,
        });

        if ai_match {
            event["shadow"] = self.gateway.maybe_shadow(&event).await;
        }
        self.log_event(event)?;

        if self.config.mitm_enabled && ai_match {
            return self
                .handle_mitm_connect(target.host, target.port, event_id, started_at, client)
                .await;
        }

        let mut upstream = match TcpStream::connect((target.host.as_str(), target.port)).await {
            Ok(stream) => stream,
            Err(error) => {
                self.log_event(json!({
                    "event": "connect_failed",
                    "host": target.host,
                    "port": target.port,
                    "error": error.kind().to_string(),
                }))?;
                return write_response(
                    &mut client,
                    502,
                    "text/plain",
                    b"upstream connect failed\n",
                    true,
                )
                .await;
            }
        };
        client
            .write_all(b"HTTP/1.1 200 Connection Established\r\n\r\n")
            .await?;
        let (bytes_out, bytes_in) = copy_bidirectional_counted(&mut client, &mut upstream).await?;
        self.log_event(json!({
            "event_id": event_id,
            "event": "connect_closed",
            "method": "CONNECT",
            "host": target.host,
            "port": target.port,
            "app": app,
            "ai_match": ai_match,
            "notion_match": notion_match,
            "duration_ms": started_at.elapsed().as_millis() as u64,
            "bytes_out": bytes_out,
            "bytes_in": bytes_in,
        }))?;
        Ok(())
    }

    async fn handle_mitm_connect(
        &self,
        host: String,
        port: u16,
        event_id: String,
        started_at: Instant,
        mut client: TcpStream,
    ) -> Result<()> {
        client
            .write_all(b"HTTP/1.1 200 Connection Established\r\n\r\n")
            .await?;
        let server_config = match server_tls_config(&host, &self.config.mitm_ca_dir) {
            Ok(config) => config,
            Err(error) => {
                self.log_event(json!({
                    "event_id": event_id,
                    "event": "payload_capture_failed",
                    "host": host,
                    "method": "CONNECT",
                    "error": error.to_string(),
                }))?;
                return Ok(());
            }
        };
        let acceptor = TlsAcceptor::from(Arc::new(server_config));
        let mut client_tls = match acceptor.accept(client).await {
            Ok(stream) => stream,
            Err(error) => {
                self.log_event(json!({
                    "event_id": event_id,
                    "event": "payload_capture_failed",
                    "host": host,
                    "method": "CONNECT",
                    "error": error.to_string(),
                }))?;
                return Ok(());
            }
        };

        let upstream_tcp = match TcpStream::connect((host.as_str(), port)).await {
            Ok(stream) => stream,
            Err(error) => {
                self.log_event(json!({
                    "event_id": event_id,
                    "event": "connect_failed",
                    "host": host,
                    "port": port,
                    "error": error.kind().to_string(),
                }))?;
                return Ok(());
            }
        };
        let connector = TlsConnector::from(Arc::new(client_tls_config()));
        let server_name =
            ServerName::try_from(host.clone()).context("invalid upstream DNS name")?;
        let mut upstream_tls = connector.connect(server_name, upstream_tcp).await?;

        let mut bytes_out = 0usize;
        let mut bytes_in = 0usize;
        loop {
            let request = match read_http_message(&mut client_tls).await {
                Ok(Some(message)) => message,
                Ok(None) => break,
                Err(error) => {
                    self.log_event(json!({
                        "event_id": event_id,
                        "event": "payload_capture_closed",
                        "host": host,
                        "error": error.to_string(),
                    }))?;
                    break;
                }
            };
            bytes_out += request.raw.len();
            let capture_event_id = Uuid::new_v4().to_string();
            let request_started_at = Utc::now();
            let request_started = Instant::now();
            let request_line = parse_request_line(&request.header_text);
            let request_payload = build_http_payload(
                &request.body,
                &request.headers,
                self.config.payload_preview_bytes,
                self.config.payload_body_bytes,
                json!({
                    "method": request_line.method,
                    "path": request_line.path,
                    "headers": redact_headers(&request.headers),
                }),
            );
            self.log_event(json!({
                "event_id": capture_event_id,
                "connection_event_id": event_id,
                "event": "http_request",
                "method": request_line.method,
                "path": request_line.path,
                "host": host,
                "app": classify_host(&host, &self.config),
                "ai_match": is_ai_host(&host, &self.config),
                "notion_match": is_notion_host(&host, &self.config),
                "headers": redact_headers(&request.headers),
                "request_bytes": request.body.len(),
                "request_preview": request_payload.get("body_preview").cloned().unwrap_or(Value::String(String::new())),
                "request_truncated": request_payload.get("preview_truncated").cloned().unwrap_or(Value::Bool(false)),
            }))?;

            upstream_tls.write_all(&request.raw).await?;
            upstream_tls.flush().await?;

            let response = match read_http_message(&mut upstream_tls).await? {
                Some(message) => message,
                None => break,
            };
            bytes_in += response.raw.len();
            let status_code = parse_response_status(&response.header_text);
            let response_payload = build_http_payload(
                &response.body,
                &response.headers,
                self.config.payload_preview_bytes,
                self.config.payload_body_bytes,
                json!({
                    "status_code": status_code,
                    "headers": redact_headers(&response.headers),
                }),
            );
            self.log_event(json!({
                "event_id": capture_event_id,
                "connection_event_id": event_id,
                "event": "http_response",
                "method": request_line.method,
                "path": request_line.path,
                "host": host,
                "app": classify_host(&host, &self.config),
                "status_code": status_code,
                "headers": redact_headers(&response.headers),
                "response_bytes": response.body.len(),
                "response_preview": response_payload.get("body_preview").cloned().unwrap_or(Value::String(String::new())),
                "response_truncated": response_payload.get("preview_truncated").cloned().unwrap_or(Value::Bool(false)),
            }))?;

            let ingest = self
                .gateway
                .ingest_capture(CaptureIngest {
                    event_id: capture_event_id.clone(),
                    host: host.clone(),
                    app: classify_host(&host, &self.config),
                    method: request_line.method.clone(),
                    path: request_line.path.clone(),
                    started_at: request_started_at,
                    ended_at: Utc::now(),
                    request_payload,
                    response_payload,
                    duration_ms: request_started.elapsed().as_millis() as u64,
                })
                .await;
            self.log_ingest(
                &capture_event_id,
                &event_id,
                &host,
                &request_line.path,
                ingest,
            )?;

            client_tls.write_all(&response.raw).await?;
            client_tls.flush().await?;

            if should_close(&request.headers) || should_close(&response.headers) {
                break;
            }
        }

        let _ = upstream_tls.shutdown().await;
        let _ = client_tls.shutdown().await;
        self.log_event(json!({
            "event_id": event_id,
            "event": "connect_closed",
            "method": "CONNECT",
            "host": host,
            "port": port,
            "app": classify_host(&host, &self.config),
            "ai_match": is_ai_host(&host, &self.config),
            "notion_match": is_notion_host(&host, &self.config),
            "capture_payloads": true,
            "duration_ms": started_at.elapsed().as_millis() as u64,
            "bytes_out": bytes_out,
            "bytes_in": bytes_in,
        }))?;
        Ok(())
    }

    async fn write_json(&self, stream: &mut TcpStream, payload: Value) -> Result<()> {
        let body = serde_json::to_vec(&payload)?;
        write_response(stream, 200, "application/json; charset=utf-8", &body, true).await
    }

    fn status_payload(&self) -> Result<Value> {
        let ca_path = if self.config.mitm_enabled {
            Some(
                ensure_ca(&self.config.mitm_ca_dir)?
                    .cert_path
                    .display()
                    .to_string(),
            )
        } else {
            None
        };
        Ok(json!({
            "listen": format!("{}:{}", self.config.host, self.config.port),
            "log_path": self.config.log_path.display().to_string(),
            "ai_domains": self.config.ai_domains,
            "notion_domains": self.config.notion_domains,
            "capture_payloads": self.config.mitm_enabled,
            "mitm_ca_path": ca_path,
            "shadow_enabled": self.config.shadow_enabled,
            "gateway_url": self.config.gateway_url,
            "events_loaded": read_events(&self.config.log_path, 1000).len(),
            "runtime": "rust",
        }))
    }

    fn log_ingest(
        &self,
        capture_event_id: &str,
        connection_event_id: &str,
        host: &str,
        path: &str,
        ingest: IngestResult,
    ) -> Result<()> {
        self.log_event(json!({
            "event_id": capture_event_id,
            "connection_event_id": connection_event_id,
            "event": "gateway_ingest",
            "host": host,
            "app": classify_host(host, &self.config),
            "path": path,
            "attempted": ingest.attempted,
            "ok": ingest.ok,
            "status": ingest.status,
            "error": ingest.error,
        }))
    }

    fn log_event(&self, event: Value) -> Result<()> {
        append_event(&self.config.log_path, event)
    }
}
